api/src/routing.js

```diff
@@ -595,4 +633,25 @@ proxyForAuditing.on('error', function(err, req, res) {
   serverUtils.serverError(JSON.stringify(err), req, res);
 });
 
+proxyForAuditing.on('proxyReq', function(proxyReq, req) {
+  // allow POST and PUT requests which have been body-parsed to be correctly proxied
+  if(req.body) {
+    let bodyData = JSON.stringify(req.body);
+    proxyReq.setHeader('Content-Type','application/json');
+    proxyReq.setHeader('Content-Length', Buffer.byteLength(bodyData));
+    proxyReq.write(bodyData);
+  }
+});
+
+// intercept responses from filtered offline endpoints to fill in with forbidden docs stubs
+proxyForAuditing.on('proxyRes', (proxyRes, req, res) => {
+  if (res.interceptResponse) {
+    let body = new Buffer('');
+    proxyRes.on('data', data => body = Buffer.concat([body, data]));
+    proxyRes.on('end', () => res.interceptResponse(body.toString()));
+  } else {
+    proxyRes.pipe(res);
+  }
+});
+
```

Comment:
It looks like unknown URLs will still end up being proxied through to couchdb even for offline users - is this right? I imagined we'd only allow online only users to access couch directly.

Reply:
Yeaaaa ... I guess I could add a flag to the `req` object before allowing it to pass through to audit, and block all offline requests which do not have this flag before audit route. Do you think that would be a good option?

Reply:
I was meaning just wrap the fallback proxying to the web in an auth check. I think it's important  that non-admins are blocked from direct access - there are just too many ways they can access data they shouldn't that we can't explicitly block them all.